Introduce lightweight E2E test for SDDC resource, using the ZEROCLOUD…

dimitarproynov opened the pull request:
… provider option

The usage of ZEROCLOUD provides the ability for the VMC Terraform provider to run an lightweight E2E test for ~100 s and zero infrastructure cost, while testing the whole SDDC lifecycle (including the create, get and delete APIs).

A subsequent change will introduce the GitHub actions yaml configuration to run the E2E acceptance test daily.

Changes:
The schema for SDDC adds missing properties, that are set by the resourceSddcRead post SDDC creation. The Terraform SDK complained about them while in acceptance test mode.

The SDDC schema properties: edrs_policy_type,  enable_edrs, min_hosts, max_hosts are now optional-computed as their value is NOT taken into account during initial SDDC creation and is computed by the VMC API. This makes any user-set values differ from what the VMC has set upon creation and running "terraform apply" on the newly created resource whould entail a change. This fails the E2E post-create validation and has been an inconvenience for our users, see:

https://github.com/vmware/terraform-provider-vmc/issues/94

Introduced the TestVMCUrl and TestCSPUrl environment variables, to enable setting the VMC and CSP service URLs for test purposes, without exposing them to the general public.

Introduces the TestAccResourceVmcSddc_Zerocloud, which is a copy of the existing TestAccResourceVmcSddc_basic, but using the ZEROCLOUD cloud provider option and a bit less checks, as the resulting SDDC is limited.

Testing done:
make build - passes
golangci-lint run - passes

Before change:
No acceptance tests ran successfully.
Right after SDDC creation running "terraform plan/apply" reported that changes are required in regard to "edrs_policy_type", "enable_edrs", "min_hosts", "max_hosts" properties, even without changes to desired configuration expressed in *.tf file.

After change:
make testacc TESTARGS="-run=TestAccResourceVmcSddc_Zerocloud" - multiple passes within 120 s

Deployed SDDC.
ran "terraform plan/apply" - no changes in regard to "edrs_policy_type", "enable_edrs", "min_hosts", "max_hosts" properties

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Introduce lightweight E2E test for SDDC resource, using the ZEROCLOUD provider option

The usage of ZEROCLOUD provides the ability for the VMC Terraform provider to run an lightweight E2E test for ~100 s and zero infrastructure cost, while testing the whole SDDC lifecycle (including the create, get and delete APIs).

A subsequent change will introduce the GitHub actions yaml configuration to run the E2E acceptance test daily.

Changes:
The schema for SDDC adds missing properties, that are set by the resourceSddcRead post SDDC creation. The Terraform SDK complained about them while in acceptance test mode.

The SDDC schema properties: edrs_policy_type,  enable_edrs, min_hosts, max_hosts are now optional-computed as their value is NOT taken into account during initial SDDC creation and is computed by the VMC API. This makes any user-set values differ from what the VMC has set upon creation and running "terraform apply" on the newly created resource whould entail a change. This fails the E2E post-create validation and has been an inconvenience for our users, see:

https://github.com/vmware/terraform-provider-vmc/issues/94

Introduced the TestVMCUrl and TestCSPUrl environment variables, to enable setting the VMC and CSP service URLs for test purposes, without exposing them to the general public.

Introduces the TestAccResourceVmcSddc_Zerocloud, which is a copy of the existing TestAccResourceVmcSddc_basic, but using the ZEROCLOUD cloud provider option and a bit less checks, as the resulting SDDC is limited.

Testing done:
make build - passes
golangci-lint run - passes

Before change:
No acceptance tests ran successfully.
Right after SDDC creation running "terraform plan/apply" reported that changes are required in regard to
"edrs_policy_type", "enable_edrs", "min_hosts", "max_hosts" properties, even without changes to desired
configuration expressed in *.tf file.

After change:
make testacc TESTARGS="-run=TestAccResourceVmcSddc_Zerocloud" - multiple passes within 120 s

Deployed SDDC.
ran "terraform plan/apply" - no changes in regard to "edrs_policy_type", "enable_edrs", "min_hosts", "max_hosts" properties

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Upgrade to Go 1.19 and golangci-lint 1.49.0 to fix error: golangci-lint exit with code 137

Upgrade of golangci-lint to 1.49.0 requires Go 1.19.
Change the usage of io/ioutils.ReadAll to io.ReadAll, as usage ioutils is deprecated.
Also removed linters structcheck, varcheck, deadcode as they are reported deprecated and replaced by "unused". see:

https://github.com/golangci/golangci-lint/issues/1841

Testing done:
golangci-lint run - passes with no warn output
make build - passes with no warn output
make testacc TESTARGS="-run=TestAccResourceVmcSddc_Zerocloud" - passes successfully

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

fstrudel commented on vmc/resource_vmc_sddc.go:
```diff
@@ -96,11 +398,22 @@ func resourceSddcCreate(d *schema.ResourceData, m interface{}) error {
+				serviceUnavailableRetries++
+				if serviceUnavailableRetries <= maxServiceUnavailableRetries {
+					return resource.RetryableError(fmt.Errorf(
+						"VMC backend is experiencing difficulties, retry to polling the SDDC Create Task"))
```
nit: I would add the current number of retry and the the max number, so this would read "Retry 5/20"

dimitarproynov replied:
Addressed

fstrudel commented on vmc/resource_vmc_sddc.go:
```diff
@@ -96,11 +398,22 @@ func resourceSddcCreate(d *schema.ResourceData, m interface{}) error {
+					return resource.RetryableError(fmt.Errorf(
+						"VMC backend is experiencing difficulties, retry to polling the SDDC Create Task"))
+				} else {
+					return resource.NonRetryableError(fmt.Errorf("VMC service is down"))
```
I would not say "VMC service is down" but more "Error: Max retries (20) reached to create SDDC"
We don't know what causing the error, VMC might be up but something else might prevent the SDDC creation.

dimitarproynov replied:
Addressed

plamzd commented on vmc/resource_vmc_sddc.go:
```diff
@@ -96,11 +398,22 @@ func resourceSddcCreate(d *schema.ResourceData, m interface{}) error {
+			// during the significant SDDC provisioning time
+			if err.Error() == (errors.ServiceUnavailable{}.Error()) {
+				serviceUnavailableRetries++
+				if serviceUnavailableRetries <= maxServiceUnavailableRetries {
+					return resource.RetryableError(fmt.Errorf(
```
Perhaps we should give it a few seconds between each retry to have a chance to recover?

dimitarproynov replied:
This code is executed inside a function provided to the RetryContext. That function is called every 10 second (the default).

Commit:
Address Fanny's review comments.

Testing done:
golangci-lint run - passes with no warn output
make build - passes with no warn output
make testacc TESTARGS="-run=TestAccResourceVmcSddc_Zerocloud" - passes successfully

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>
